File: 1/2021.11.03/407.Trapping Rain Water II/python/solution.py
class Solution:
    def trapRainWater(self, heightMap: List[List[int]]) -> int:
        if len(heightMap) <= 2 or len(heightMap[0]) <= 2:
            return 0
        m, n = len(heightMap), len(heightMap[0])
        heap = []
        for j in range(n):
            heappush(heap, (heightMap[0][j], j))
            heightMap[0][j] = -1
            heappush(heap, (heightMap[m - 1][j], (m - 1) * n + j))
            heightMap[m - 1][j] = -1
        for i in range(1, m - 1):
            heappush(heap, (heightMap[i][0], i * n))
            heightMap[i][0] = -1
            heappush(heap, (heightMap[i][n - 1], i * n + n - 1))
            heightMap[i][n - 1] = -1
        res = 0
        dirs = [-1, 0, 1, 0, -1]
        while heap:
            height, cur = heappop(heap)
            for k in range(4):
                nx, ny = cur // n + dirs[k], cur % n + dirs[k + 1]
                if 0 <= nx < m and 0 <= ny < n and heightMap[nx][ny] != -1:
                    if height > heightMap[nx][ny]:
                        res += height - heightMap[nx][ny]
                    heappush(heap, (max(height, heightMap[nx][ny]), nx * n + ny))
                    heightMap[nx][ny] = -1
        return res

File: 1/2021.11.03/407.Trapping Rain Water II/java/solution.java
class Solution {
    public int trapRainWater(int[][] heightMap) {
        if (heightMap.length <= 2 || heightMap[0].length <= 2) {
            return 0;
        }
        int m = heightMap.length, n = heightMap[0].length;
        boolean[][] visited = new boolean[m][n];
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (i == 0 || i == m - 1 || j == 0 || j == n - 1) {
                    heap.offer(new int[]{i * n + j, heightMap[i][j]});
                    visited[i][j] = true;
                }
            }
        }
        int res = 0;
        int[] dirs = {-1, 0, 1, 0, -1};
        while (!heap.isEmpty()) {
            int[] cur = heap.poll();
            for (int k = 0; k < 4; ++k) {
                int nx = cur[0] / n + dirs[k];
                int ny = cur[0] % n + dirs[k + 1];
                if (nx >= 0 && nx < m && ny >= 0 && ny < n && !visited[nx][ny]) {
                    if (cur[1] > heightMap[nx][ny]) {
                        res += cur[1] - heightMap[nx][ny];
                        visited[nx][ny] = true;
                    }
                    heap.offer(new int[]{nx * n + ny, Math.max(heightMap[nx][ny], cur[1])});
                    visited[nx][ny] = true;
                }
            }
        }
        return res;
    }
}
